**Simone Rittenhouse**

**Assignment 1**

"Edinburgh_Buses.db" has two tables:

"route" containing:

1.   the route number ("num")
2.   the bus company name ("company")
3.   the position of the stop given by "stop" along the route ("pos")
4.   the stop number ("stop")

"stops" containing:

1.   the stop number ("id")
2.   "stop name"

Two companies can use the same route number, but those would be considered different routes.



In [ ]:
# code executed in Google Colab
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
gdrivePath="/content/drive/MyDrive/"
import sqlite3
import pandas as pd
# create a connection to a database
conn = sqlite3.connect(gdrivePath+"Edinburgh_Buses.db")

In [ ]:
def execSQL(conn,query):
  conn.execute(query) # execute an SQL query
  conn.commit() # "commit" that query in order to make its action permanent

In [ ]:
def allrowsSelect(conn,query):
  cursor = conn.execute(query)
  for row in cursor:
   print(row)

1. Write a query to find all of the distinct bus companies.

In [ ]:
query = 'SELECT DISTINCT company FROM route'

allrowsSelect(conn, query)

('LRT',)
('MAC',)
('SMT',)
('SMJ',)
('CCC',)
('MID',)
('LOW',)
('LRJ',)
('WIL',)
('FS',)
('FF',)
('JM',)
('ROU',)
('VL',)


2. Write a query that lists all of the distinct route number and company combinations.

In [ ]:
query = 'SELECT DISTINCT num, company FROM route'

allrowsSelect(conn, query)

('1', 'LRT')
('10', 'LRT')
('100', 'LRT')
('100', 'MAC')
('106', 'SMT')
('11', 'LRT')
('113', 'SMT')
('12', 'LRT')
('124', 'SMT')
('129', 'SMT')
('13', 'LRT')
('134', 'SMT')
('137', 'SMT')
('14', 'LRT')
('142', 'SMJ')
('15', 'LRT')
('16', 'LRT')
('16', 'SMT')
('17', 'LRT')
('18', 'SMT')
('19', 'LRT')
('19', 'SMT')
('19A', 'LRT')
('2', 'LRT')
('20', 'LRT')
('21', 'LRT')
('22', 'LRT')
('22A', 'LRT')
('23', 'LRT')
('24', 'LRT')
('25', 'LRT')
('251', 'CCC')
('26', 'LRT')
('26A', 'LRT')
('27', 'LRT')
('27', 'MID')
('275', 'SMT')
('28', 'LRT')
('29', 'LOW')
('2A', 'LRT')
('3', 'LRT')
('30', 'LOW')
('30', 'LRT')
('31', 'LRT')
('315', 'LOW')
('32', 'LRT')
('33', 'LRT')
('34', 'LRT')
('35', 'LRT')
('36', 'LRT')
('37', 'LRT')
('37', 'SMT')
('38', 'LRT')
('38', 'MID')
('3A', 'LRJ')
('4', 'LRT')
('40', 'LRT')
('41', 'LRT')
('42', 'LRT')
('43', 'LRT')
('43', 'SMT')
('44', 'LRT')
('45', 'LRT')
('455', 'MID')
('46A', 'LRT')
('47', 'LRT')
('47', 'MID')
('48', 'WIL')
('49', 'SMT')
('5', 'LRT')
('51', '

3. Find a query to find the route numbers that are used by more than one company, and the number of times that they are used.

In [ ]:
query = '''
SELECT num, COUNT(DISTINCT company) as count
FROM route
GROUP BY num
HAVING count > 1
'''

allrowsSelect(conn, query)

('100', 2)
('16', 2)
('19', 2)
('27', 2)
('30', 2)
('37', 2)
('38', 2)
('43', 2)
('47', 2)
('55', 2)
('61', 2)
('65', 2)
('80', 2)
('85', 2)
('86', 2)
('87', 2)


4. Write a query to find the names of all of the stops on route 17 operated by company LRT, ordered by position. Output the position, stop number, and stop name.

In [ ]:
query = '''
SELECT r.pos, r.stop, s.name
FROM route as r INNER JOIN stops as s ON r.stop = s.id
WHERE r.num == 17 AND r.company == 'LRT'
ORDER BY r.pos
'''

allrowsSelect(conn, query)

(1, 167, 'Newhaven')
(2, 137, 'Leith')
(3, 141, 'Leith Walk')
(4, 194, 'Princes Street')
(5, 230, 'Tollcross')
(6, 160, 'Morningside')
(7, 179, 'Oxgangs')


5. Write a query that counts how many distinct routes each company has. It should output the company name and the number of routes.

In [ ]:
query = '''
SELECT company, COUNT(DISTINCT num) as routes
FROM route
GROUP BY company
'''

allrowsSelect(conn, query)

('CCC', 1)
('FF', 1)
('FS', 2)
('JM', 1)
('LOW', 6)
('LRJ', 1)
('LRT', 72)
('MAC', 1)
('MID', 4)
('ROU', 2)
('SMJ', 1)
('SMT', 38)
('VL', 1)
('WIL', 1)


6. Write a query to find all of the distinct routes operated by company "LOW".

In [ ]:
query = '''
SELECT DISTINCT num as routes
FROM route
WHERE company == 'LOW'
'''

allrowsSelect(conn, query)

('29',)
('30',)
('315',)
('62',)
('94',)
('95',)


7. Find the ten routes with the most stops. Remember that a route is identified by both its company and number.

In [ ]:
query = '''
SELECT num, company, COUNT(stop) as count
FROM route
GROUP BY num, company
ORDER BY count DESC
LIMIT 10
'''

allrowsSelect(conn, query)

('32', 'LRT', 23)
('52', 'LRT', 23)
('12', 'LRT', 18)
('2A', 'LRT', 18)
('2', 'LRT', 17)
('34', 'LRT', 15)
('35', 'LRT', 15)
('66A', 'SMT', 13)
('81A', 'LRT', 13)
('14', 'LRT', 12)


8. Find the ten routes with the fewest stops.

In [ ]:
query = '''
SELECT num, company, COUNT(stop) as count
FROM route
GROUP BY num, company
ORDER BY count
LIMIT 10
'''

allrowsSelect(conn, query)

('60', 'JM', 2)
('59', 'LRT', 3)
('100', 'LRT', 4)
('48', 'WIL', 4)
('55', 'LRT', 4)
('86A', 'LRT', 4)
('93', 'SMT', 4)
('VL2', 'VL', 4)
('49', 'SMT', 5)
('58', 'LRT', 5)


9. Do any routes have the same stop on them more than once? Write a query to check if they do, by creating a table containing the num, company, and stop columns for all of the stops that appear more than one, along with the number of times that they appear.

In [ ]:
# creating table (view)
query='''
CREATE VIEW same_stops AS
SELECT num, company, stop, COUNT(stop) as stop_count
FROM route
GROUP BY num, company, stop
HAVING stop_count > 1
'''
execSQL(conn,query)

# querying new table (view)
query='''
SELECT * from same_stops
'''
allrowsSelect(conn, query)

('1', 'LRT', 137, 2)
('12', 'LRT', 31, 2)
('12', 'LRT', 115, 2)
('12', 'LRT', 173, 2)
('14', 'LRT', 215, 2)
('19', 'LRT', 28, 2)
('19A', 'LRT', 28, 2)
('2', 'LRT', 31, 2)
('2', 'LRT', 115, 2)
('2', 'LRT', 173, 2)
('275', 'SMT', 48, 2)
('275', 'SMT', 165, 2)
('2A', 'LRT', 31, 2)
('2A', 'LRT', 115, 2)
('2A', 'LRT', 173, 2)
('32', 'LRT', 173, 2)
('34', 'LRT', 92, 2)
('34', 'LRT', 213, 2)
('35', 'LRT', 92, 2)
('35', 'LRT', 213, 2)
('42', 'LRT', 36, 2)
('46A', 'LRT', 36, 2)
('52', 'LRT', 173, 2)
('6', 'LRT', 137, 2)
('81', 'LRT', 47, 2)
('81', 'LRT', 59, 2)
('81', 'LRT', 162, 2)
('81', 'LRT', 219, 2)
('81A', 'LRT', 47, 2)
('81A', 'LRT', 59, 2)
('81A', 'LRT', 162, 2)
('81A', 'LRT', 219, 2)
('R1', 'ROU', 106, 2)
('R2', 'ROU', 106, 2)


10. Find the numbers and names of the ten stops that are on the most routes, along with the number of routes they are on. Count a stop only once for each route it appears on. For instance, stop 137 appears both at the beginning (position 1) and the end (position 11) of route 1 of company LRT; stop 137 should be counted only once for this route.

In [ ]:
query = '''
SELECT stop, name, COUNT(stop) as stop_count
FROM (SELECT DISTINCT r.stop as stop, s.name as name, r.num, r.company 
      FROM route as r INNER JOIN stops as s ON r.stop = s.id)
GROUP BY stop
ORDER BY stop_count DESC
LIMIT 10
'''

allrowsSelect(conn, query)

(219, 'St Andrew Square', 47)
(115, 'Haymarket', 42)
(194, 'Princes Street', 38)
(29, 'Bridges', 36)
(48, 'Corstorphine', 27)
(149, 'London Road', 25)
(101, 'Gorgie', 20)
(137, 'Leith', 20)
(163, 'Musselburgh', 20)
(213, 'Sighthill', 20)


11. Find all of the routes that go from stop 219 to stop 115. The position of stop 219 on the route needs to be before that of 115.

In [ ]:
query = '''
SELECT num, company 
FROM (SELECT a.stop, b.stop, a.num as num, a.company as company
      FROM route as a INNER JOIN route as b ON a.num = b.num AND a.company = b.company
      WHERE a.stop == 219 AND b.stop == 115)
'''

allrowsSelect(conn, query)

('137', 'SMT')
('22', 'LRT')
('22A', 'LRT')
('25', 'LRT')
('37', 'SMT')
('38', 'MID')
('61', 'LRT')
('61', 'SMT')
('63', 'LRT')
('C12', 'SMT')
